arroyo-sql/src/tables.rs

```diff
@@ -0,0 +1,616 @@
+                )
+                .collect(),
+        };
+
+        let physical_schema = DFSchema::new_with_metadata(
+            physical_struct
+                .fields
+                .iter()
+                .map(|f| {
+                    let TypeDef::DataType(data_type, nullable ) = f.data_type.clone() else {
+                        bail!("expect data type for generated column")
+                    };
+                    Ok(DFField::new_unqualified(&f.name, data_type, nullable))
+                })
+                .collect::<Result<Vec<_>>>()?,
+            HashMap::new(),
+        )?;
+
+        let expression_context = ExpressionContext {
+            input_struct: &physical_struct,
+            schema_provider,
+        };
+
+        let sql_to_rel = SqlToRel::new(schema_provider);
+        struct_field_pairs
+            .into_iter()
+            .map(|(mut struct_field, generating_expression)| {
+                if let Some(generating_expression) = generating_expression {
+                    // TODO: Implement automatic type coercion here, as we have elsewhere.
+                    // It is done by calling the Analyzer which inserts CAST operators where necessary.
+
+                    let df_expr = sql_to_rel.sql_to_expr(
+                        generating_expression,
+                        &physical_schema,
+                        &mut PlannerContext::default(),
+                    )?;
+                    let expr = expression_context.compile_expr(&df_expr)?;
+                    struct_field.expression = Some(Box::new(expr));
+                }
+
+                Ok(struct_field)
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+
+    pub fn try_from_statement(
+        statement: &Statement,
+        schema_provider: &ArroyoSchemaProvider,
+    ) -> Result<Option<Self>> {
+        if let Statement::CreateTable {
+            name,
+            columns,
+            with_options,
+            query: None,
+            ..
+        } = statement
+        {
+            let name: String = name.to_string();
+            let mut with_map = HashMap::new();
+            for option in with_options {
+                with_map.insert(
+                    option.name.value.to_string(),
+                    value_to_inner_string(&option.value)?,
+                );
+            }
+
+            let fields = Self::schema_from_columns(columns, schema_provider)?;
+
+            let connector = with_map.remove("connector");
+
+            match connector.as_ref().map(|c| c.as_str()) {
+                Some("memory") | None => {
+                    if fields.iter().any(|f| f.expression.is_some()) {
+                        bail!("Virtual fields are not supported in memory tables; instead write a query");
+                    }
+
+                    if !with_map.is_empty() {
+                        bail!("Memory tables do not allow any with options; to create a connector table set the 'connector' option");
```

Comment:
This triggers for a query like
```
CREATE TABLE bid_count (
  bids bigint
) WITH (
    connector = 'memory',
    format = 'json'
);
INSERT INTO bid_count
SELECT count(*) as bids FROM small_nexmark where bid is not null;
SELECT * FROM bid_count
```
And the error message is confusing
> Memory tables do not allow any with options; to create a connector table set the 'connector' option

Reply:
I think you suggested supporting `connector = 'memory'` — do you have a suggestion for how to handle this?

Reply:
I updated it to show different error messages for the `connector = 'memory'` case and the no-connector case